h2/src/main/org/h2/mvstore/db/MVSecondaryIndex.java

```diff
@@ -252,6 +252,27 @@ public void remove(Session session, Row row) {
         }
     }
 
+    @Override
+    public void update(Session session, Row oldRow, Row newRow) {
+        if (!rowsAreEqual(oldRow, newRow)) {
+            super.update(session, oldRow, newRow);
+        }
+    }
+
+    private boolean rowsAreEqual(SearchRow rowOne, SearchRow rowTwo) {
+        if (rowOne == rowTwo) {
+            return true;
+        }
+        for (int index : columnIds) {
+            Value v1 = rowOne.getValue(index);
+            Value v2 = rowTwo.getValue(index);
+            if (v1 == null ? v2 != null : !v1.equals(v2)) {
+                break;
```

Comment:
`return false`?

BTW, `ValueStringIgnoreCase.equals()` unlike other value types may return `true` for different strings. But this ugly data type can be addressed in a separate issue, it may have more unrelated problems due to its `equals()` implementation.

Reply:
I filled a #1216 for `ValueStringIgnoreCase`.

Reply:
I have only one complain about incorrect `break` statement here.

Reply:
`equals(…)` is just fine for `NUMERIC` / `DECIMAL` and `TIMESTAMP WITH TIME ZONE`, it always returns `false` if values are somehow different. This is a valid situation in Java, objects with `compareTo(…) == 0` may have `equals(…) == false` and we simply should use proper methods depending on situation.

But case-insensitive strings are not sane now.

Reply:
Yes, hash()/equals() contract is broken, not the end of the world, because it's not a hash table. As long as equals is consistent with comparison... Initially I was using comparison, and it is ok to skip index update if values are compare returns 0 (even if equals() gives false). Problem might appear later when index key will be used by select to avoid access to primary index, and that's where we need a flag for those bad/partially ordered types saying "don't do it, fetch from the table (primary index)"

Reply:
`hashCode()` is actually used in `ValueCache` but it does not matter right now, `ValueCache` does not detect such problems. The actual problem is that `ValueStringIgnoreCase.equals()` may return `true` for different values that is not expected not only by indexes, but also by some commands. I sent a test case for such issue with `UPDATE` command and columns with `ON UPDATE`.